File: androidApp/src/main/kotlin/io/umain/munchies/android/analytics/FirebaseAnalyticsService.kt
package io.umain.munchies.android.analytics

import io.umain.munchies.core.analytics.AnalyticsEvent
import io.umain.munchies.core.analytics.AnalyticsService
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class FirebaseAnalyticsService(
    private val dispatcher: CoroutineDispatcher = Dispatchers.IO
) : AnalyticsService {

    private val firebaseAnalytics by lazy {
        try {
            val firebaseClass = Class.forName("com.google.firebase.ktx.Firebase")
            val analyticsProperty = firebaseClass.getMethod("getAnalytics")
            analyticsProperty.invoke(null)
        } catch (e: Exception) {
            null
        }
    }

    override suspend fun trackEvent(event: AnalyticsEvent) {
        withContext(dispatcher) {
            val analytics = firebaseAnalytics ?: return@withContext
            val bundle = android.os.Bundle().apply {
                when (event) {
                    is AnalyticsEvent.ScreenView -> {
                        putString("screen_name", event.screenName)
                        putString("screen_class", event.screenClass)
                        event.previousScreen?.let { putString("previous_screen", it) }
                    }

                    is AnalyticsEvent.TabSwitch -> {
                        putString("tab_id", event.tabId)
                        putString("tab_name", event.tabName)
                    }

                    is AnalyticsEvent.ModalOpen -> {
                        putString("modal_name", event.modalName)
                        putString("modal_class", event.modalClass)
                    }

                    is AnalyticsEvent.ModalDismiss -> {
                        putString("modal_name", event.modalName)
                        putLong("time_spent_ms", event.timeSpentMs)
                    }

                    is AnalyticsEvent.ErrorEvent -> {
                        putString("error_type", event.errorType)
                        putString("error_message", event.errorMessage)
                        putString("error_context", event.errorContext)
                    }

                    is AnalyticsEvent.CustomEvent -> {}
                }

                event.properties.forEach { (key, value) ->
                    runCatching { putString(key, value) }
                }
            }

            runCatching {
                analytics::class.java.getMethod("logEvent", String::class.java, android.os.Bundle::class.java)
                    .invoke(analytics, event.eventName, bundle)
            }
        }
    }

    override suspend fun setUserProperties(properties: Map<String, String>) {
        withContext(dispatcher) {
            val analytics = firebaseAnalytics ?: return@withContext
            properties.forEach { (key, value) ->
                runCatching {
                    analytics::class.java.getMethod("setUserProperty", String::class.java, String::class.java)
                        .invoke(analytics, key, value)
                }
            }
        }
    }

    override suspend fun clearUserProperties() {
        withContext(dispatcher) {
            val analytics = firebaseAnalytics ?: return@withContext
            runCatching {
                analytics::class.java.getMethod("setUserId", String::class.java)
                    .invoke(analytics, null)
            }
        }
    }

    override suspend fun flush() {
        withContext(dispatcher) {}
    }
}

File: iosApp/iosApp/Analytics/FirebaseAnalyticsService.swift
import Foundation
import FirebaseAnalytics
import shared

class FirebaseAnalyticsService: NSObject, AnalyticsService {
    private let queue = DispatchQueue(label: "io.umain.analytics.ios", qos: .background)
    
    func trackEvent(event: AnalyticsEvent) async throws {
        await queue.async {
            let bundle = NSMutableDictionary()
            
            switch event {
            case let screenView as AnalyticsEvent.ScreenView:
                bundle["screen_name"] = screenView.screenName
                bundle["screen_class"] = screenView.screenClass
                if let previousScreen = screenView.previousScreen {
                    bundle["previous_screen"] = previousScreen
                }
                
            case let tabSwitch as AnalyticsEvent.TabSwitch:
                bundle["tab_id"] = tabSwitch.tabId
                bundle["tab_name"] = tabSwitch.tabName
                
            case let modalOpen as AnalyticsEvent.ModalOpen:
                bundle["modal_name"] = modalOpen.modalName
                bundle["modal_class"] = modalOpen.modalClass
                
            case let modalDismiss as AnalyticsEvent.ModalDismiss:
                bundle["modal_name"] = modalDismiss.modalName
                bundle["time_spent_ms"] = NSNumber(value: modalDismiss.timeSpentMs)
                
            case let errorEvent as AnalyticsEvent.ErrorEvent:
                bundle["error_type"] = errorEvent.errorType
                bundle["error_message"] = errorEvent.errorMessage
                bundle["error_context"] = errorEvent.errorContext
                
            case is AnalyticsEvent.CustomEvent:
                break
                
            default:
                break
            }
            
            event.properties.forEach { (key, value) in
                bundle[key] = value
            }
            
            Analytics.logEvent(event.eventName, parameters: bundle as? [String: Any])
        }
    }
    
    func setUserProperties(properties: [String : String]) async throws {
        await queue.async {
            properties.forEach { (key, value) in
                Analytics.setUserProperty(value, forName: key)
            }
        }
    }
    
    func clearUserProperties() async throws {
        await queue.async {
            Analytics.setUserID(nil)
        }
    }
    
    func flush() async throws {
        await queue.async {
            Analytics.resetAnalyticsData()
        }
    }
}
